File: CPP/Offer/58/ReverseStringII.cpp
// 
// @ClassName ReverseStringII
// @Description Leetcode 541  Reverse String II
// @Date 18-10-21 下午10:22
// @Created by Insomnia
//

#include <string>
#include <iostream>

using namespace std;

class Solution {
public:
    string reverseString(string s, int n) {
        reverseStr(s, 0, n - 1);
        reverseStr(s, n, s.length() - 1);
        reverseStr(s, 0, s.length() - 1);
        return s;
    }

private:
    void reverseStr(string &s, int l, int r) {
        while (l < r) {
            swap(s[l++], s[r--]);
        }
    }
};

int main() {
    string s = "abcdefg";

    Solution solution;
    cout << solution.reverseString(s, 2) << endl;

    return 0;
}

File: Python3/Offer/58/ReverseStringII.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# @Time    : 2019/1/15 11:21 AM
# @Author  : Insomnia
# @Desc    : 翻转部分字符串例如, abcdefg, 翻转成, defgabc
# @File    : ReverseStringII.py
# @Software: PyCharm

# 此实现方法和c++完全不同
class Solution:
    def reverPartStr(self, str, idx):
        str1 = str[:idx]
        str2 = str[idx:]
        str3 = str2+str1
        return str3
    def __reverseStr(self, str):
        return str[::-1]

if __name__ == '__main__':
    str1 = "abcdefg"
    str2 = "1234567"
    sol = Solution()
    print(sol.reverPartStr(str1, 2))
    print(sol.reverPartStr(str2, 4))
